### DISCUSSION TITLE
Fuzzy search gives rather wrong results in entity picker

### DISCUSSION THREAD

**Discussion by ildar170975** [home-assistant/frontend#17474] [Dashboard]
After adding a fuzzy search logic into entity picker, it gives lots of wrong results.
Here is an example - other 3 entities have maximum "`su`" substring:

![image](https://github.com/home-assistant/frontend/assets/71872483/de3c2b08-c038-4ef8-bcc1-7de2cfd55919)

I can understand intentions of people who asked about adding this fuzzy logic, but the current implementation gives rather wrong results.

**Comment by breakthestatic** [comment]
This appears correct. As noted in my other comments, the fuzzy search will give the highest rank to the **closest** match. In this case, `sun.sun` matches the query `sun.`, and the "closeness" of each of the constituent characters (adjacent) means that it will rank very highly, and therefore appear at the top of the list.

However, fuzzy search essentially matches any results that have the query characters, separated by any number of *other* characters, as long as they are in order.  In this example, it's a bit tough to say what that match is, since the entities are cut-off.

I definitely see "sun" in both searched fields:
entity_id:  **s**ensor.iiyama_as**u**s_p8z68_v_pro_**n**uvoton_nct6776.
name: IIYAMA A**SU**S P8Z68-V PRO **N**uvoton NCT6776F Volta

There are the ellipses for the overflow, but those are purely visual and shouldn't be included in the match. Can you provide the **full** strings for both the `entity_id` and `name` fields?  If there's no period/dot in the remainder of either field, then this is actually a bug.

I did try to recreate it locally by using a dummy entity with what I can see of your names/ids. `entity_id` can't have a `.`, so that pretty much means the `name` field is likely where the `.` could be. When I include a `.` in the entity name, it *does* appear in the list, which is correct, since IIYAMA A**SU**S P8Z68-V PRO **N**uvoton NCT6776F Voltage<b>.</b>Test now matches. Conversely, if I remove the `.` in the entity name, only `sun.sun` appears, since there's no longer any `.` **after** the `n` match in the string.
![Screenshot 2023-08-03 at 9 31 52 AM](https://github.com/home-assistant/frontend/assets/9346314/ab6eb951-610d-471e-b68d-363d462e4be0)
![Screenshot 2023-08-03 at 9 59 33 AM](https://github.com/home-assistant/frontend/assets/9346314/9dd67670-d6b8-4997-9e55-48695063f6b8)

**Comment by ildar170975** [comment]
> Can you provide the **full** strings for both the `entity_id` and `name` fields? I

Here it is:
```
sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_voltages_cpu_vcore
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Voltages CPU VCore

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_voltages_12v
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Voltages +12V

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_voltages_analog_3_3v
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Voltages Analog +3.3V

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_voltages_3_3v
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Voltages +3.3V

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_voltages_5v
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Voltages +5V

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_voltages_standby_3_3v
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Voltages Standby +3.3V

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_voltages_vbat
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Voltages VBAT

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_temperatures_cpu
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Temperatures CPU

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_temperatures_auxiliary
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Temperatures Auxiliary

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_temperatures_motherboard
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Temperatures Motherboard

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_fans_fan_1
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Fans Fan #1

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_fans_fan_2
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Fans Fan #2

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_controls_fan_1
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Controls Fan #1

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_controls_fan_2
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Controls Fan #2

sensor.iiyama_asus_p8z68_v_pro_nuvoton_nct6776f_controls_fan_3
IIYAMA ASUS P8Z68-V PRO Nuvoton NCT6776F Controls Fan #3
```

**Reply by ildar170975** [reply]
Sorry, I cannot comment now the algorithm you are using here, my brains are really far away from any math this moment.
But IMHO "fuzzy logic" means these scenarios: if "`sun.`" is defined then these entries are selected:
- sun.
- abc sun.abc
- abcsun .abc
- .sun
- abc. sun

i.e. criteria "abc xyz" will give us
- abcxyz
- abc kjlkjlkjlkjj xyz
- abcklklklklklyxyz
- xyz abc
- xyzabc
- xyz jhjhjhjh abc
- xyzjkhkjhkhkhabc

IMHO the search criteria should be split into parts like "`sun`" & "`.`", not into parts "`s`", "`u`" etc.

In a perfect world we should have a kind of "closeness" setting for the in-built search in the current picker (default is "exact", a user may change it to get more results).

**Comment by breakthestatic** [comment]
Ahh that explains it... the ones that are appearing when you query "sun." **do** have dots in their names (specifically the three 3.3V ones):

All of the following match fuzzy search because of the dots:

<sup>IIYAMA A</sup>SU<sup>S P8Z68-V PRO </sup>N<sup>uvoton NCT6776F Voltages Analog +3</sup>.<sup>3V</sup>

<sup>IIYAMA A</sup>SU<sup>S P8Z68-V PRO </sup>N<sup>uvoton NCT6776F Voltages +3</sup>.<sup>3V</sup>

<sup>IIYAMA A</sup>SU<sup>S P8Z68-V PRO </sup>N<sup>uvoton NCT6776F Voltages Standby +3</sup>.<sup>3V</sup>

The fuzzy search utility, as implemented by Home Assistant, allows **any** number of other characters between the matching ones. The only real criteria is that all of the characters in the query string exist, **in order** in the target. There are other algorithms that take proximity into account in the scoring, but that isn't how the pre-existing filter/sort used by HA is configured.

BTW, not like it helps, but I didn't code the fuzzy search at all. It already existed and was used in other parts of the UI, just not the "picker" components yet.  I do agree that some type of closeness setting would be ideal, but @karwosts indicated such a setting would likely not fly with the lead devs.  I'd try looking into adding that, but am not very familiar with any of the settings persistence aspects of HA, and given the low likelihood of being merged, would be a somewhat large time-sink. Perhaps someone more well-versed could take a look though?

**Reply by karwosts** [reply]
I think if someone wanted to go look through the fuzzyScore algorithm a good case for a PR could be made to set the filter threshold at something other than "undefined", someone would just have to do a test with some cases and pick a threshold that seems to return sane choices.  

Agree that just having all the characters is pretty weak filter.

**Reply by ildar170975** [reply]
For me a more nice algorithm will be:
-- split to lexemes (i.e. parts divided by a whitespace or characters like ".", "-", "_" ...)
-- combine these lexemes in ANY order (lexeme1 lexeme2 lexeme3, lexeme3 lexeme2 lexeme1, ...)
These gives us entities "Peter room", "Peter room on 2nd floor", "Room of Peter", "peter_room", "peter_dining_room" ...

**Comment by ildar170975** [comment]
> The only real criteria is that all of the characters in the query string exist

Well, this explains that strange behaviour.
The criteria is split into CHARACTERS, not into "lexemes".
Not saying that this is critical for me; but inconvenient. Just IMHO.

**Comment by ildar170975** [comment]
The latest changes give us this in the Entities list (just a small fragment of a huge list of entities):

![image](https://github.com/home-assistant/frontend/assets/71872483/24ae3d14-5f38-4899-8594-68d3d3924154)

In Dev tools -> set state it gives us this:

![image](https://github.com/home-assistant/frontend/assets/71872483/ba001905-c280-4e27-bea8-1c27fcfc9923)

and if "sun." is inputted:

![image](https://github.com/home-assistant/frontend/assets/71872483/b734d7a6-08af-4194-ad45-33451f56b189)

**Comment by AndreKR** [comment]
Winamp had the perfect search decades ago, I don't understand why nowadays this seems to be such a hard problem.

Split the search term at whitespace into parts, then show all items that contain all the parts, anywhere in any of their labels, keeping whatever sort order the items have.
